### PR TITLE
cache: dedup prefetch requests to avoid read amplification

### PR DESCRIPTION
**Pull request by polyrabbit**
We observed a large read amplification when user is doing an incremental but non-continuous sequential read on large amount of files. After labeling object GET requests with purpose, we found prefetching is the main cause of the read amplification.

It turns out the deduplication in `prefetcher::do` method does not work as expected especially when there is only one prefetcher (the default behavior), which caused duplicated prefetching. Instead, it's better to do deduplication before inserting into the `pending` queue.

image

### CODE REVIEW

File: pkg/chunk/prefetch.go [hunk 1]
```diff
@@ -39,23 +39,23 @@ func newPrefetcher(parallel int, fetch func(string)) *prefetcher {
...
-			delete(p.busy, key)
-		}
+		p.Lock()
+		delete(p.busy, key)
 		p.Unlock()
 	}
 }
 
 func (p *prefetcher) fetch(key string) {
+	p.Lock()
```
Comment (jiefenghuang) [inline pkg/chunk/prefetch.go:51]: this may be blocked and fetch should not block the read process

Comment (polyrabbit) [reply pkg/chunk/prefetch.go:51]: Which line it will be blocked on?

**Comment by jiefenghuang** [comment]
The unit tests in "prefetch_test.go" also need to be changed.

**Comment by polyrabbit** [comment]
@jiefenghuang Sorry I missed your last message. Which part in prefetch_test.go do you think needs to be changed?

**Pull request merged**